### PR TITLE
ENH: special: use boost in `nctdtr`

### LINKED ISSUE #8344: BUG: special.nctdtr: incorrect results
Reported to me by @jb-leger who had weird issue with `scipy.stats.nct.logcdf`, and tracked down a bit more.

### Reproducing code example:
```
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

import scipy.special as sc 
x = np.linspace(-1,1, 100)
df=3000
nc=3
y = sc.nctdtr(df, nc, x)
plt.plot(x,y)
```

download-3

### Error message:

No errors. Plot just seem wrong with a step which should not be here. Or at least should raise an error message if parameters are outside supported range.

### Scipy/Numpy/Python version information:
```
SciPy: 1.1.0.dev0+beb258e  (Jan 24 2018)
NumPy:  1.13.3 
Python:  sys.version_info(major=3, minor=6, micro=3, releaselevel='final', serial=0)
```

I believe that end up calling `scipy/special/cdflib/cumtnc.f`

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Closes https://github.com/scipy/scipy/issues/8344

#### What does this implement/fix?
Uses boost for `nctdtr`. The PR is modeled after #21505. The trickiest part was the mpmath reference implementation,.

#### Additional information
I did not migrate the noncentral t quantile `nctdtrit` yet as all default tests (no xslow etc.) ran successfully locally and I have to adjust the size of my PRs to the sleeping cycles of my little one, so trying to keep the small.

### CODE REVIEW

File: scipy/special/tests/test_cdflib.py [hunk 1]
```diff
@@ -585,3 +584,50 @@ def test_ncfdtr(dfn, dfd, nc, f, expected):
     # sample_idx = rng.choice(len(re), replace=False, size=12)
     # cases = np.array(cases)[sample_idx].tolist()
     assert_allclose(sp.ncfdtr(dfn, dfd, nc, f), expected, rtol=1e-13, atol=0)
+
+
+@pytest.mark.parametrize(
+    "df,nc,x,expected,rtol",
+    [[3000., 3., 0.1, 0.0018657780826323328, 1e-13],
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py:591]: This is the test of the parameter range in the linked issue.

File: scipy/special/boost_special_functions.h [hunk 2]
```diff
@@ -951,26 +951,46 @@ ncf_kurtosis_excess_double(double v1, double v2, double l)
...
+    if (std::isnan(x) || std::isnan(v) || std::isnan(l)) {
+	return NAN;
     }
-    // -inf => 0, inf => 1
-    return 1.0 - std::signbit(x);
+    if (v <= 0) {
+	sf_error("nctdtr", SF_ERROR_DOMAIN, NULL);
+	return NAN;
+    }
+    if (std::isinf(x)) {
```
Comment (fancidev) [inline scipy/special/boost_special_functions.h:963]: `std::isinf(x)` is `true` even if `x` is negative infinity. In this case should return 0?

File: scipy/special/boost_special_functions.h [hunk 3]
```diff
@@ -951,26 +951,46 @@ ncf_kurtosis_excess_double(double v1, double v2, double l)
...
+	return 1.0;
+    }
+    Real y;
+    try {
+	y = boost::math::cdf(
+                boost::math::non_central_t_distribution<Real, SpecialPolicy>(v, l), x);
```
Comment (fancidev) [inline scipy/special/boost_special_functions.h:969]: I guess this is a more general question than being specific to this PR, but would it be possible to tailor `SpecialPolicy` to align with `scipy.special`’s conventions and  eliminate the need to `catch (...)` at each invocation?

File: scipy/special/boost_special_functions.h [hunk 4]
```diff
@@ -951,26 +951,46 @@ ncf_kurtosis_excess_double(double v1, double v2, double l)
...
+    } catch (...) {
+	/* Boost was unable to produce a result. */
+        sf_error("nctdtr", SF_ERROR_NO_RESULT, NULL);
+        y = NAN;
+    }
+    if ((y < 0) || (y > 1)) {
```
Comment (fancidev) [inline scipy/special/boost_special_functions.h:975]: I think this is a reasonable check. I wonder if you could find some input to trigger this branch of the code, if only for coverage purpose?

An alternative “trick” is to clamp the return value within $[0,1]$, which I see being used occasionally. Of course this relies on the assumption that the return value is out of range due to (small) numerical noise instead of wild algorithmic error.

**Review by fancidev — COMMENTED** [review]
Looks good to me! Just some small comments below.

File: scipy/special/boost_special_functions.h (hunk 2 above)
Comment (dschmitz89) [reply scipy/special/boost_special_functions.h:963]: Yep, good point. What is a good way to check for neg. infinity in C++ ?

File: scipy/special/boost_special_functions.h (hunk 4 above)
Comment (steppi) [reply scipy/special/boost_special_functions.h:975]: We don’t clamp because we’ve observed Boost returning complete nonsense far out of range in some cases. See the discussion for #21505.

File: scipy/special/boost_special_functions.h (hunk 3 above)
Comment (steppi) [reply scipy/special/boost_special_functions.h:969]: We catch here and call `sf_error` with the correct error code so that error handling can be controlled by [special.errstate](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.errstate.html) and [special.seterr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.seterr.html#scipy.special.seterr).

File: scipy/special/boost_special_functions.h [hunk 5]
```diff
@@ -951,26 +951,46 @@ ncf_kurtosis_excess_double(double v1, double v2, double l)
...
+	return NAN;
     }
-    // -inf => 0, inf => 1
-    return 1.0 - std::signbit(x);
+    if (v <= 0) {
+	sf_error("nctdtr", SF_ERROR_DOMAIN, NULL);
+	return NAN;
+    }
+    if (std::isinf(x)) {
+	return 1.0;
```
Comment (steppi) [inline scipy/special/boost_special_functions.h:964]:
```suggestion
    if (std::isinf(x)) {
	return  (x > 0) ? 1.0 : 0.0;
```

File: scipy/special/boost_special_functions.h (hunk 2 above)
Comment (fancidev) [reply scipy/special/boost_special_functions.h:963]: Can do `x == -INF`, or testing for `x < 0` inside the `isinf(x)` branch.

File: scipy/special/boost_special_functions.h (hunk 3 above)
Comment (fancidev) [reply scipy/special/boost_special_functions.h:969]: Indeed I see `ncf_ppf_wrap` of https://github.com/scipy/scipy/pull/21505/files properly “translates” `boost` exceptions to `special` error codes.

File: scipy/special/boost_special_functions.h (hunk 4 above)
Comment (dschmitz89) [reply scipy/special/boost_special_functions.h:975]: I did not stress test this function as much as @steppi did with the noncentral F distribution. I am quite sure though that there must be dark corners as the NCT and NCF distributions are based on the same non central beta in boost. Both of them also heavily rely on the incomplete beta function which can go bananas for extreme input.

**Comment by dschmitz89** [comment]
@josef-pkt : would migrating `nctdtr` to boost potentially cause issues for statsmodels? Like with `ncfdtr`, for the vast majority of cases we get much better precision. In case you have roundtrip tests for `nctdtrit` or `nctdtrinc`, they might fail for extreme inputs though.

**Comment by josef-pkt** [comment]
I switched away from scipy stats distributions for both nct and ncf to use scipy special
https://github.com/statsmodels/statsmodels/issues/8659

AFAIR, AFAICS for nct it was only a runtime warning that showed up during rootfinding in sample size computation. 
It did not affect the final results. So, I don't know which corner/extreme case evaluated during the rootfinding caused the Runtime Warning.

(Because I always use older versions of scipy, lagging 1 to 3 years behind, I cannot investigate new problems when they show up.)

**Comment by dschmitz89** [comment]
> I switched away from scipy stats distributions for both nct and ncf to use scipy special [statsmodels/statsmodels#8659](https://github.com/statsmodels/statsmodels/issues/8659)
> 
> AFAIR, AFAICS for nct it was only a runtime warning that showed up during rootfinding in sample size computation. It did not affect the final results. So, I don't know which corner/extreme case evaluated during the rootfinding caused the Runtime Warning.
> 
> (Because I always use older versions of scipy, lagging 1 to 3 years behind, I cannot investigate new problems when they show up.)

Looking at https://github.com/statsmodels/statsmodels/issues/8624, it would be interesting to know which parameters caused Boost's noncentral t CDF to go wild.

**Comment by josef-pkt** [comment]
I have now scipy 0.10

```
>>> stats.nct.cdf(4.30265273, 2., 0.978343)
Traceback (most recent call last):
  File "<pyshell#15>", line 1, in <module>
    stats.nct.cdf(4.30265273, 2., 0.978343)
  File "...\WPy64-31140\python-3.11.4.amd64\Lib\site-packages\scipy\stats\_distn_infrastructure.py", line 2186, in cdf
    place(output, cond, self._cdf(*goodargs))
  File "...\WPy64-31140\python-3.11.4.amd64\Lib\site-packages\scipy\stats\_continuous_distns.py", line 6826, in _cdf
    return np.clip(_boost._nct_cdf(x, df, nc), 0, 1)
RuntimeWarning: invalid value encountered in _nct_cdf
```

this is just looking at the values in the debugger of reported case in statsmodels issue
(I change the statsmodels code to use scipy.stats nct.cdf)

**Comment by bashtage** [comment]
I don't recall the specifics TBH.  It was failing in pre-release testing which led to the previous bug report.

**Comment by dschmitz89** [comment]
> stats.nct.cdf(4.30265273, 2., 0.978343)

Both warnings do not appear anymore with a recent scipy version. Maybe you could just suppress it.

This PR is ready from my side.

**Comment by josef-pkt** [comment]
Thanks,
We can keep using the scipy special version of nct in statsmodels for now, then users that upgrade to the next scipy will get the fixed nct boost version.

File: scipy/special/boost_special_functions.h [hunk 6]
```diff
@@ -951,26 +951,46 @@ ncf_kurtosis_excess_double(double v1, double v2, double l)
...
 {
-    if (std::isfinite(x)) {
-        return boost::math::cdf(
-            boost::math::non_central_t_distribution<Real, StatsPolicy>(v, l), x);
+    if (std::isnan(x) || std::isnan(v) || std::isnan(l)) {
+	return NAN;
     }
-    // -inf => 0, inf => 1
-    return 1.0 - std::signbit(x);
+    if (v <= 0) {
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h:959]: Actually `v=0` is valid, we would get the regular f distribution CDF, right?

**Comment by dschmitz89** [comment]
@steppi: if you could provide an example/best practice of how to use the new rootfinder with boost functions, I could tackle migrating a lot of cdflib functions over to boost over the next months piece by piece. Such PRs are not really popular as they are tedious but they fit my current time budget for individual PRs reasonably well.

File: scipy/special/tests/test_cdflib.py [hunk 7]
```diff
@@ -461,40 +460,6 @@ def test_chndtrix_gh2158():
                82.35640899964173, 84.45263768373256]
     assert_allclose(res, res_exp)
 
-@pytest.mark.xfail_on_32bit("32bit fails due to algorithm threshold")
```
Comment (steppi) [inline scipy/special/tests/test_cdflib.py:464]: Why are we removing these test cases? We're they failing because the old cdflib reference is inaccurate for some of them? The reference values could be replaced with `mpmath` then. This seems like a good opportunity to remove the `xfail_on_32bit` and show that they are still passing.

Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py:464]: Yes, the old reference values were very inaccurate for some cases. Allright, I will integrate them in the new tests against the mpmath reference. I might make a subjective selection though if that's okay with you.

Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py:464]: For high noncentrality values of 38 such as in the test originally, the mpmath reference computes CDF values of the order of `e-310`. There also boost fails to get close or returns nan, so I dropped these for noncentrality of 15 instead.

Comment (steppi) [reply scipy/special/tests/test_cdflib.py:464]: Cases like this where the reference output is subnormal are why it makes sense to use an `atol` in tests. In cases like this, it's fine to return zero and signal underflow through `sf_error`. Is Boost returning `nan` where the previous method used to return 0? I'd count that as a bug.

Comment (bashtage) [reply scipy/special/tests/test_cdflib.py:464]: Fully agree that the `nan` feels like a bug.

**Comment by steppi** [comment]
> @steppi: if you could provide an example of how to use the new rootfinder with boost functions, I could tackle migrating a lot of cdflib functions over to boost over the next months piece by piece. Such PRs are not really popular as they are tedious but they fit my current time budget for individual PRs reasonably well.

Sure, that would be great. I'll do one when I get bandwidth for it, and will ping you on the PR. This would likely be after the [xsf migration](https://github.com/scipy/xsf/issues/1), which I hope to complete before we branch.

File: scipy/special/tests/test_cdflib.py (hunk 7 above)
Comment (steppi) [reply scipy/special/tests/test_cdflib.py:464]: @dschmitz89, do you plan to follow up on these `nan`s?

Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py:464]: @steppi : I do , am just short on time at the moment. Not sure I can make it till the 1.15 release as I am also going on holiday at the end of the month. I want to spend my scipy activities until then mostly on review. If you can/want to take over the last todos here, that would be much appreciated as I think this PR is almost ready to merge.

Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py:464]: @steppi : I just had another look at the test cases for `nc=38`. Even for small values within the normal floating point range, we get high discrepancies.

For `df, nc, t = 0.98, 38, 1.5` the mpmath reference computes `1.546906350750292e-295` but boost results in `2.5919953604832226e-97`. Would you prefer to get the test to pass with something like `atol=1e-90`? I would simply ignore these edge cases for now to be honest. In the parameter ranges that matter most we get much better results with boost.

Comment (steppi) [reply scipy/special/tests/test_cdflib.py:464]: It looks like you set `mp.dps` too low to evaluate the sum. I set `mp.dps=500` and got the result `2.591995360483094e-97` from your reference implementation. Much closer to Boost. If you're evaluating a sum with `mpmath`, you may need to set `dps` pretty high.

Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py:464]: Ouch, thanks for spotting this. With `mp.dps=500`, the reference calculation did not finish within 10 minutes on my laptop though. It has performance issues at the moment but this is crazy. How long does it take on your machine? Might try out my Mac instead.

Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py:464]: With 400 significant digits I also get `2.591995360483094e-97` within 12 minutes of computation time.

**Comment by dschmitz89** [comment]
Allright, stars aligned and I had a bit of time yesterday.

The `nan` value originates from `nctdtr(980, 38, 1.5)`, here boost computes a tiny value of app. `-1e-70`. In the wrapper, this is set to `nan`. I worked around it in the test by ignoring it. We could theoretically also clip for tiny negative values to zero but who knows if that is the better option for other cases.

There is now one failing parameter combination for 32bit CPUs. Let's skip the test again on 32bit then?

@steppi : let me know if you have a strong opinion here about any of the above points, otherwise I will add the test skip.

**Comment by steppi** [comment]
@dschmitz89, instead of removing the failing test case, mark it as `xfail` with a reason, and then do a conditional `xfail` for only the one failing case on 32bit.

Syntax is, instead of having a test case

`[0.98, -3.8, 0.15, 0.9999528361700505, 1e-13],` in the `pytest.mark`parametrize` list, put

````python
            pytest.param(
                [0.98, -3.8, 0.15, 0.9999528361700505, 1e-13],
                marks=pytest.mark.xfail(reason="Bug in underlying Boost math implementation"),
            )
```

for the failing test case on 32bit, your `pytest.param` will look like

```python
pytest.param(
    test_case,
    marks=pytest.mark.xfail(condition=sys.maxsize < 2**32, reason="Fails on 32 bit.")
)
```

I don't think it's a good idea to just get rid of failing test cases, or skip on 32bit because of only one failing case.

File: scipy/special/tests/test_cdflib.py [hunk 8]
```diff
@@ -592,50 +592,81 @@ class TestNctdtr:
...
+        (980, 3.8, 1.5, 0.010750678886113882),
+        (980, 3.8, 15, 1.0),
+        (980, 38, 0.0015, 3.0547506e-316),
+        (980, 38, 0.15, 8.6191646313e-314),
+        pytest.param(980, 38, 1.5, 1.1824454111413493e-291,
+                     marks=pytest.mark.xfail(
+                        reason="Bug in underlying Boost math implementation")),
+        (980, 38, 15, 5.407535300713606e-105)
+    ])
+    def test_nctdtr_gh19896(self, df, nc, x, expected):
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py:663]:
```suggestion
    def test_gh19896(self, df, nc, x, expected):
```

File: scipy/special/tests/test_cdflib.py [hunk 9]
```diff
@@ -585,3 +550,140 @@ def test_ncfdtr(dfn, dfd, nc, f, expected):
...
+        # test that gh-19896 is resolved.
+        # Originally this was a regression test that used the old Fortran results
+        # as a reference. The Fortran results were not accurate, so the reference
+        # values were recomputed with mpmath.
+        result = sp.nctdtr(df, nc, x)
+        assert_allclose(result, expected, rtol=1e-13, atol=1e-303)
+
+
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py:671]:
```suggestion
```

File: scipy/special/tests/test_cdflib.py [hunk 10]
```diff
@@ -585,3 +550,140 @@ def test_ncfdtr(dfn, dfd, nc, f, expected):
...
+    #         return f
+
+    #     if x >= 0:
+    #         result = f(df, nc, x)
+    #     else:
+    #         result = mp.one - f(df, -nc, x)
+    #     return float(result)
+
+    @pytest.mark.parametrize("df, nc, x, expected", [
+        [0.98, -3.8, 0.0015, 0.9999279987514815],
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py:596]:
```suggestion
        (0.98, -3.8, 0.0015, 0.9999279987514815),
```

File: scipy/special/tests/test_cdflib.py [hunk 11]
```diff
@@ -585,3 +550,140 @@ def test_ncfdtr(dfn, dfd, nc, f, expected):
...
+        "df, nc, x, expected, rtol",
+        [[3., 5., -2., 1.5645373999149622e-09, 5e-9],
+         [1000., 10., 1., 1.1493552133826623e-19, 1e-13],
+         [1e-5, -6., 2., 0.9999999990135003, 1e-13],
+         [10., 20., 0.15, 6.426530505957303e-88, 1e-13],
+         [1., 1., np.inf, 1.0, 0.0],
+         [1., 1., -np.inf, 0.0, 0.0]
+        ]
+    )
+    def test_nctdtr_accuracy(self, df, nc, x, expected, rtol):
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py:688]:
```suggestion
    def test_accuracy(self, df, nc, x, expected, rtol):
```

**Comment by dschmitz89** [comment]
Interesting that the 32bit job succeeds now. Maybe it had an issue in the way I handled the nan case which is now xfailed. Will push the latest stylistic updates without rerunning CI.

**Comment by dschmitz89** [comment]
The green CI results do not show up here anymore due to my last commit: for the reviewer, they can easily be found [here](https://github.com/scipy/scipy/actions?query=actor%3Adschmitz89).
